### PR TITLE
[cuda] masked_scatter : static_cast init_value to circumvent cuda 11.2 issue

### LINKED ISSUE #51544: Several masked_scatter tests in test_autograd failed with Jacobian mismatch
## 🐛 Bug

Several `masked_scatter` test failures in `test_autograd` found on 2070, 3090, V100, A100 with cuda 11.2

Some commits happened on January 27th may cause this problem. Although I'm not certain, https://github.com/pytorch/pytorch/pull/50039 looks suspicious. 

## To Reproduce

As of commit 7328710cbce459bbd77dfe424acc94010725d979, run this

```
$ python test/test_autograd.py -v -k test_masked_scatter_broadcast_lhs_cuda 
Test results will be stored in test-reports/python-unittest

Running tests...
----------------------------------------------------------------------
  test_masked_scatter_broadcast_lhs_cuda (__main__.TestAutogradDeviceTypeCUDA) ... ERROR (1.581s)

======================================================================
ERROR [1.581s]: test_masked_scatter_broadcast_lhs_cuda (__main__.TestAutogradDeviceTypeCUDA)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/xwang/Developer/pytorch/torch/testing/_internal/common_utils.py", line 872, in wrapper
    method(*args, **kwargs)
  File "/home/xwang/Developer/pytorch/torch/testing/_internal/common_utils.py", line 872, in wrapper
    method(*args, **kwargs)
  File "/home/xwang/Developer/pytorch/torch/testing/_internal/common_device_type.py", line 295, in instantiated_test
    raise rte
  File "/home/xwang/Developer/pytorch/torch/testing/_internal/common_device_type.py", line 290, in instantiated_test
    result = test_fn(self, *args)
  File "/home/xwang/Developer/pytorch/test/test_autograd.py", line 5189, in do_test
    check(name)
  File "/home/xwang/Developer/pytorch/test/test_autograd.py", line 5110, in check
    run_functional_checks(self, test_name, name, fn,
  File "/home/xwang/Developer/pytorch/test/test_autograd.py", line 4982, in run_functional_checks
    run_grad_and_gradgrad_checks(test_case, name, test_name, apply_fn,
  File "/home/xwang/Developer/pytorch/test/test_autograd.py", line 4960, in run_grad_and_gradgrad_checks
    test_case.assertTrue(gradcheck(apply_method, input_variables, eps=1e-6, atol=PRECISION,
  File "/home/xwang/Developer/pytorch/torch/testing/_internal/common_utils.py", line 1909, in gradcheck
    return torch.autograd.gradcheck(fn, inputs, **kwargs)
  File "/home/xwang/Developer/pytorch/torch/autograd/gradcheck.py", line 468, in gradcheck
    checkIfNumericalAnalyticAreClose(a, n, j)
  File "/home/xwang/Developer/pytorch/torch/autograd/gradcheck.py", line 449, in checkIfNumericalAnalyticAreClose
    return fail_test(error_str + 'Jacobian mismatch for output %d with respect to input %d,\n'
  File "/home/xwang/Developer/pytorch/torch/autograd/gradcheck.py", line 367, in fail_test
    raise RuntimeError(msg)
RuntimeError: Jacobian mismatch for output 0 with respect to input 1,
numerical:tensor([[1.0000, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
        [0.0000, 0.0000, 0.0000,  ..., 1.0000, 0.0000, 0.0000],
        [0.0000, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
        ...,
        [0.0000, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
        [0.0000, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
        [0.0000, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000]],
       device='cuda:0')
analytical:tensor([[1., 0., 0.,  ..., 0., 0., 0.],
        [0., 0., 0.,  ..., 0., 0., 0.],
        [0., 0., 0.,  ..., 0., 0., 0.],
        ...,
        [0., 0., 0.,  ..., 0., 0., 0.],
        [0., 0., 0.,  ..., 0., 0., 0.],
        [0., 0., 0.,  ..., 0., 0., 0.]], device='cuda:0')


----------------------------------------------------------------------
Ran 1 test in 1.581s

FAILED (errors=1)

Generating XML reports...
```

However, there were also several other `masked_scatter` tests failed in `test_autograd` (similar Jacobian mismatch) at January 28th night df07e1c, but these tests somehow disappeared as of the current commit.

At df07e1c, failed tests were

```
ERROR [0.121s]: test_masked_scatter_broadcast_lhs_cuda (__main__.TestAutogradDeviceTypeCUDA)
ERROR [0.154s]: test_masked_scatter_broadcast_rhs_cuda (__main__.TestAutogradDeviceTypeCUDA)
ERROR [0.146s]: test_masked_scatter_cuda (__main__.TestAutogradDeviceTypeCUDA)
ERROR [0.150s]: test_masked_scatter_scalar_broadcast_rhs_cuda (__main__.TestAutogradDeviceTypeCUDA)
ERROR [0.247s]: test_masked_select_broadcast_all_cuda (__main__.TestAutogradDeviceTypeCUDA)
ERROR [0.061s]: test_masked_select_broadcast_lhs_cuda (__main__.TestAutogradDeviceTypeCUDA)
ERROR [0.091s]: test_masked_select_broadcast_rhs_cuda (__main__.TestAutogradDeviceTypeCUDA)
ERROR [0.087s]: test_masked_select_cuda (__main__.TestAutogradDeviceTypeCUDA)
ERROR [0.056s]: test_masked_select_scalar_broadcast_lhs_cuda (__main__.TestAutogradDeviceTypeCUDA)
ERROR [0.107s]: test_masked_select_scalar_broadcast_rhs_cuda (__main__.TestAutogradDeviceTypeCUDA)
```

For example, as of commit 7328710cbce459bbd77dfe424acc94010725d979, I run

```
$ python test/test_autograd.py -v -k test_masked_scatter_cuda               
Test results will be stored in test-reports/python-unittest

Running tests...
----------------------------------------------------------------------

----------------------------------------------------------------------
Ran 0 tests in 0.000s

OK

Generating XML reports...
```

Only the 1 of 10 tests still remains at this moment, or, the other 9 were fixed and this one was left alone. 🤔 

## Expected behavior

No fail.

## Environment

My 2070

```
Collecting environment information...
PyTorch version: 1.8.0a0+7328710
Is debug build: False
CUDA used to build PyTorch: 11.2
ROCM used to build PyTorch: N/A

OS: Manjaro Linux (x86_64)
GCC version: (GCC) 10.2.0
Clang version: Could not collect
CMake version: version 3.19.2

Python version: 3.9 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: GeForce RTX 2070 SUPER
GPU 1: GeForce GTX 1070 Ti

Nvidia driver version: 455.45.01
cuDNN version: Probably one of the following:
/usr/lib/libcudnn.so.8.0.5
/usr/lib/libcudnn_adv_infer.so.8.0.5
/usr/lib/libcudnn_adv_train.so.8.0.5
/usr/lib/libcudnn_cnn_infer.so.8.0.5
/usr/lib/libcudnn_cnn_train.so.8.0.5
/usr/lib/libcudnn_ops_infer.so.8.0.5
/usr/lib/libcudnn_ops_train.so.8.0.5
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.5
[pip3] torch==1.8.0a0+unknown
[conda] Could not collect
```

Also seeing failures on Nvidia DGX A100, DGX V100, RTX 3090 with cuda 11.2

## Additional context

Can someone also check if Circle CI detected this failure?

cc @ezyang @anjali411 @dylanbespalko @mruberry @ngimel @ptrblck @kshitij12345

### PR DESCRIPTION
**Pull request by kshitij12345**
Fixes https://github.com/pytorch/pytorch/issues/51544

Tested locally as there is no CI for 11.2 as of now.

### CODE REVIEW

**Review by ngimel — APPROVED** [review]

File: aten/src/ATen/native/cuda/IndexKernel.cu [hunk 1]
```diff
@@ -324,14 +324,24 @@ void masked_scatter_cuda_impl(Tensor& self, const Tensor& mask, const Tensor& so
   // number of elements available in `src`
   TORCH_CHECK(totalElements <= srcSize, "source nElements must be == mask `1` elements");
 
+
+#if CUDA_VERSION >= 1120
```
Comment (xwang233) [inline aten/src/ATen/native/cuda/IndexKernel.cu:328]: For rocm compatibility, we may change this to 

```cpp
#if defined(CUDA_VERSION) && CUDA_VERSION >= 11020
...
```

File: aten/src/ATen/native/cuda/IndexKernel.cu [hunk 2]
```diff
@@ -324,14 +324,24 @@ void masked_scatter_cuda_impl(Tensor& self, const Tensor& mask, const Tensor& so
...
+  // FIXME: there appears to be a bug in Thrust (CUDA 11.2) for mixed
+  // iterator prefix sums? Convert `mask` to the same datatype as what
+  // we're accumulating the prefix sum in (int64_t) to get around it
+  // Reference: https://github.com/pytorch/pytorch/issues/51544
+  auto mask_cont = mask.to(kLong).contiguous();
+  thrust::device_ptr<int64_t> maskData(mask_cont.data_ptr<int64_t>());
+#elif
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/IndexKernel.cu:335]: hip build error here is real, you probably need `else`

Comment (kshitij12345) [reply aten/src/ATen/native/cuda/IndexKernel.cu]: Yup. Just saw that.
Updated.
Thanks!

File: aten/src/ATen/native/cuda/IndexKernel.cu (hunk 1 above)
Comment (kshitij12345) [reply aten/src/ATen/native/cuda/IndexKernel.cu:328]: Sure! Thanks!

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**